File: packages/mirascript/src/vm/lib/global/sequence/unique.ts
import { toBoolean } from '../../../../helpers/convert/to-boolean.js';
import { $Call } from '../../../operations/call.js';
import { isSame } from '../../../operations/utils.js';
import type { VmAny, VmConst, VmValue } from '../../../types/index.js';
import { expectArray, expectCallable, VmLib } from '../../helpers.js';

/** 默认相等 */
function defaultEqual(a: VmValue = null, b: VmValue = null): boolean {
    return isSame(a, b);
}

/** 获取相等函数 */
function eq(equaler: VmAny, recovered: VmValue): typeof defaultEqual {
    if (equaler == null) return defaultEqual;
    expectCallable('equal', equaler, recovered);
    return (a: VmValue = null, b: VmValue = null) => {
        const ret = $Call(equaler, [a, b]);
        return toBoolean(ret, undefined);
    };
}

export const unique = VmLib(
    (data, equal) => {
        expectArray('data', data, null);
        const e = eq(equal, data);
        const arr: VmConst[] = [];
        for (const v of data) {
            let found = false;
            for (const u of arr) {
                if (e(v, u)) {
                    found = true;
                    break;
                }
            }
            if (!found) arr.push(v ?? null);
        }
        return arr;
    },
    {
        summary: '对数组中的元素进行去重，并返回去重后的结果',
        params: {
            data: '要去重的数组',
            equal: '用于判等两个元素的函数，返回一个布尔值，默认使用严格相等比较',
        },
        paramsType: {
            data: 'array',
            equal: 'fn(a: any, b: any) -> boolean',
        },
        returnsType: 'array',
        examples: ['unique([1, 2, 2, 3]) // [1, 2, 3]'],
    },
);

export const unique_by = VmLib(
    (data, key, equal) => {
        expectArray('data', data, null);
        expectCallable('key', key, data);
        const e = eq(equal, data);
        const arr: VmConst[] = [];
        const keys: VmValue[] = [];
        const len = data.length;
        for (let i = 0; i < len; i++) {
            const v = data[i] ?? null;
            const k = $Call(key, [v, i, data]);
            let found = false;
            const keysLen = keys.length;
            for (let j = 0; j < keysLen; j++) {
                if (e(k, keys[j])) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                arr.push(v);
                keys.push(k);
            }
        }
        return arr;
    },
    {
        summary: '根据键函数对数组中的元素进行去重，并返回去重后的结果',
        params: {
            data: '要去重的数组',
            key: '用于提取去重键的函数，接受一个元素并返回其去重键',
            equal: '用于判同两个元素的函数，返回一个布尔值，默认使用严格相等比较',
        },
        paramsType: {
            data: 'array',
            key: 'fn(value: any, index: number, arr: type(data)) -> any',
            equal: 'fn(a: any, b: any) -> boolean',
        },
        returnsType: 'array',
        examples: ['unique_by(["apple", "banana", "apricot"], fn { chars(it)[0] }) // ["apple", "banana"]'],
    },
);

File: crates/python/mirascript/vm/lib/vm_global/sequence/unique.py
from .....helpers.convert.to_boolean import toBoolean
from ....operations import Call, is_same
from ....types import VmValue
from ..._helpers import _expect_array, _expect_callable
from ....types.types import Uninitialized


def default_equal(a=None, b=None):
    if a is Uninitialized:
        a = None
    if b is Uninitialized:
        b = None
    return is_same(a, b)


def eq(equaler, recovered):
    if equaler is None or equaler is Uninitialized:
        return default_equal
    _expect_callable("equal", equaler, recovered)

    def equal(a: VmValue, b: VmValue):
        ret = Call(equaler, a, b)
        return toBoolean(ret)

    return equal


def unique(data=Uninitialized, equal=Uninitialized):
    _expect_array("data", data, None)
    e = eq(equal, data)
    arr = []
    for item in data:
        found = False
        for unique_item in arr:
            if e(item, unique_item):
                found = True
                break
        if not found:
            arr.append(item)
    return arr


def unique_by(data=Uninitialized, key_fn=Uninitialized, equal=Uninitialized):
    _expect_array("data", data, None)
    _expect_callable("key_fn", key_fn, data)
    e = eq(equal, data)

    arr = []
    keys = []

    l = len(data)

    for i in range(l):
        item = data[i]
        key = Call(key_fn, item, i, data)
        found = False
        for unique_key in keys:
            if e(key, unique_key):
                found = True
                break
        if not found:
            arr.append(item)
            keys.append(key)
    return arr
